### PR TITLE
Cli app to generate password file entry.

### PR DESCRIPTION
**Pull request by bdpiparva**
* -h or help to print usage

To generate passwordfile entry using bcrypt execute following command:

```
$ java -jar gopasswd.jar -a bcrypt -u <username-here> -p <password-here>
```

If -p indicates, Use the password from the command line rather than prompting for it.

Checkout usage for more information.

### CODE REVIEW

File: gocdpasswd/src/main/java/cd/go/authentication/crypt/PasswordGenerator.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            }
+
+            CliArguments cliArguments = CliArguments.fromArgs(args);
+
+            final String hash = cliArguments.algorithm().hashProvider().hash(cliArguments);
+
+            System.out.println("\n" + hash + "\n");
+
+        } catch (Exception e) {
+            System.err.println(e.getMessage());
```
Comment (ketan) [inline gocdpasswd/src/main/java/cd/go/authentication/crypt/PasswordGenerator.java]: Fail with non zero exit code?

File: gocdpasswd/src/main/java/cd/go/authentication/crypt/PasswordReader.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            String reTypePassword = new String(console.readPassword("Re-type new password: "));
+
+            if (newPassword.equals(reTypePassword)) {
+                return newPassword;
+            }
+        } finally {
+            console.flush();
+        }
+
+        throw new RuntimeException("Password verification error.");
```
Comment (ketan) [inline gocdpasswd/src/main/java/cd/go/authentication/crypt/PasswordReader.java]: What about blank passwords? It's that valid?

File: gocdpasswd/src/main/java/cd/go/authentication/crypt/hash/BCryptProvider.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import cd.go.authentication.crypt.CliArguments;
+import org.mindrot.jbcrypt.BCrypt;
+
+import static java.text.MessageFormat.format;
+
+public class BCryptProvider implements HashProvider {
+
+    @Override
+    public String hash(CliArguments arguments) {
+        if (arguments.cost() > 15) {
```
Comment (ketan) [inline gocdpasswd/src/main/java/cd/go/authentication/crypt/hash/BCryptProvider.java]: This is arbitrary. On a slow computer, even an argument of 5 may make forever. Instead, just say the statement without the surrounding if clause.

Comment (ketan) [inline gocdpasswd/src/main/java/cd/go/authentication/crypt/hash/BCryptProvider.java]: Additionally, the client should not print anything unless there is an error. Use case being using this in an automated tool like chef, in which case htpasswd's behavior and cli are good.

**Comment by ketan** [comment]
The PR looks good apart from compatibility issues with htpasswd's cli and output format.

File: gocdpasswd/src/main/java/cd/go/authentication/crypt/PasswordReader.java (hunk 2 above)
Comment (bdpiparva) [inline gocdpasswd/src/main/java/cd/go/authentication/crypt/PasswordReader.java]: Yes, it is valid. you can generate hash for blank password using `htpasswd`.

**Comment by ketan** [comment]
@bdpiparva See http://jcommander.org/#_password instead of echoing a password on the console

**Comment by ketan** [comment]
Ignore me, I see it's not being echoed.

**Pull request merged**